### DISCUSSION TITLE
Error with AzureAD

### DISCUSSION THREAD

**Discussion by Mike-E-angelo** [datalust/seq-tickets#2199] [Help] [answered]
After using the latest Seq image I see the following error being thrown after attempting to login via Azure Directory:

```
System.InvalidOperationException: IDX20803: Unable to obtain configuration from: 'https://login.microsoftonline.com//v2.0/.well-known/openid-configuration'. Will retry at '05/22/2024 15:19:18 +00:00'. Exception: 'System.IO.IOException: IDX20807: Unable to retrieve document from: '[PII of type 'System.String' is hidden. For more details, see https://aka.ms/IdentityModel/PII.]'. HttpResponseMessage: '[PII of type 'System.Net.Http.HttpResponseMessage' is hidden. For more details, see https://aka.ms/IdentityModel/PII.]', HttpResponseMessage.Content: '[PII of type 'System.String' is hidden. For more details, see https://aka.ms/IdentityModel/PII.]'.\n   at Microsoft.IdentityModel.Protocols.HttpDocumentRetriever.GetDocumentAsync(String address, CancellationToken cancel)\n   at Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfigurationRetriever.GetAsync(String address, IDocumentRetriever retriever, CancellationToken cancel)\n   at Microsoft.IdentityModel.Protocols.ConfigurationManager`1.GetConfigurationAsync(CancellationToken cancel)'.\n ---> System.IO.IOException: IDX20807: Unable to retrieve document from: '[PII of type 'System.String' is hidden. For more details, see https://aka.ms/IdentityModel/PII.]'. HttpResponseMessage: '[PII of type 'System.Net.Http.HttpResponseMessage' is hidden. For more details, see https://aka.ms/IdentityModel/PII.]', HttpResponseMessage.Content: '[PII of type 'System.String' is hidden. For more details, see https://aka.ms/IdentityModel/PII.]'.\n   at Microsoft.IdentityModel.Protocols.HttpDocumentRetriever.GetDocumentAsync(String address, CancellationToken cancel)\n   at Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfigurationRetriever.GetAsync(String address, IDocumentRetriever retriever, CancellationToken cancel)\n   at Microsoft.IdentityModel.Protocols.ConfigurationManager`1.GetConfigurationAsync(CancellationToken cancel)\n   --- End of inner exception stack trace ---\n   at Microsoft.IdentityModel.Protocols.ConfigurationManager`1.GetConfigurationAsync(CancellationToken cancel)\n   at Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.HandleChallengeAsyncInternal(AuthenticationProperties properties)\n   at Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.HandleChallengeAsync(AuthenticationProperties properties)\n   at Microsoft.AspNetCore.Authentication.AuthenticationHandler`1.ChallengeAsync(AuthenticationProperties properties)\n   at Microsoft.AspNetCore.Authentication.AuthenticationService.ChallengeAsync(HttpContext context, String scheme, AuthenticationProperties properties)\n   at Microsoft.AspNetCore.Mvc.ChallengeResult.ExecuteResultAsync(ActionContext context)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|30_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeResultFilters>g__Awaited|28_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)\n   at Seq.Server.Web.Middleware.WebSocketAcceptMiddleware.Invoke(HttpContext context)\n   at Seq.Server.Web.Middleware.RequestAuthenticationMiddleware.Invoke(HttpContext httpContext)\n   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)\n   at Seq.Server.Web.Middleware.CanonicalUriMiddleware.Invoke(HttpContext context)\n   at Seq.Server.Web.Middleware.ServerStatusMiddleware.Invoke(HttpContext context)\n   at Seq.Server.Web.Middleware.UniversalHeadersMiddleware.Invoke(HttpContext context)\n   at Seq.Server.Web.Middleware.RequestCompletionMiddleware.Invoke(HttpContext httpContext)","RequestMethod":"GET","RequestPath":"/oidc/challenge"
```

I haven't changed any code around this and this worked with previous images.  Is there a new setting to consider?  I see the double `//` in the URL (`https://login.microsoftonline.com//v2.0/.well-known/openid-configuration`) and wonder if this is causing the issue? 🤔  Thank you for any assistance you can provide.

**Comment by nblumhardt** [comment]
Hi Mike, thanks for your message.

We haven't heard of any similar issues from other customers (a lot use Entra ID), and this version works for one of our long-running dogfood environments, so unfortunately it's not an obvious issue.

I looked into the `//` thing (nice spotting) but it seems this is within the realm of "normal" for AAD/OIDC:

https://learn.microsoft.com/en-us/entra/identity-platform/scopes-oidc#trailing-slash-and-default

I'll dig in further and let you know what I find. I think we may have a setting squirreled away to turn on PII in exception messages, and if so that might yield up some clues.

At your end, I'd recommend looking into the container's network configuration, making sure the Seq image has outbound access to that address. The URI in question is one of the "back plane" ones, requested by Seq itself from the AAD authority.

**Reply by nblumhardt** [reply]
Hi @Mike-E-angelo,

I've had a look into this; I don't think the `//` is the one mentioned in that link above - I think it's where your "tenant id" should be included in the URL.

If I include another token, there, I see:

![image](https://github.com/datalust/seq-tickets/assets/342712/9d768c02-7c3a-45b5-82c6-bca4a91d4bb3)

This matches the docs:

https://learn.microsoft.com/en-us/entra/identity-platform/v2-protocols-oidc#find-your-apps-openid-configuration-document-uri

So then I guess the question becomes - how could your tenant id be missing here?

And that got me back to remembering that you're automating deployment of Seq; is this a freshly-deployed instance? The `AzureADTenantId` setting (and all of the AAD settings) were renamed a version or two back to `EntraID*`; existing settings would have been migrated, but if you're running `seqcli setting set` using an older `seqcli` the call may have failed silently and this could have gone unnoticed before switching over the auth provider.

Am I on the right track? HTH!

**Reply by Mike-E-angelo** [reply]
Ah!  That totally must be it then, @nblumhardt!  Thank you for pointing that out.  Did `AuthenticationProvider` change as well?  This is what I have:

```
        seqcli setting set -n=EntraIDAuthority -v="login.microsoftonline.com" -s $Address --apikey=$system
        seqcli setting set -n=EntraIDClientId -v="$application" -s $Address --apikey=$system
        seqcli setting set -n=EntraIDClientKey -v="$secret" -s $Address --apikey=$system
        seqcli setting set -n=EntraIDTenantId -v="$TenantId" -s $Address --apikey=$system
        seqcli setting set -n=AuthenticationProvider -v="Azure Active Directory" -s $Address --apikey=$system
```

**Reply by Mike-E-angelo** [reply]
Unfortunately even with the above I still get `//` in the requesting URI. 🤔  Is there an announcement/documentation regarding this change?  I can further review it if so.  I did a search and didn't see anything obvious.

EDIT: switching the image from `datalust/seq:latest` to `datalust/seq:2024.1.11028` fixes this issue for now.

**Reply by nblumhardt** [reply]
Hi Mike, thanks for the updates. Instead of `"Azure Active Directory"` the `AuthenticationProvider` setting is now `"Microsoft Entra ID"`.

The ticket linked from the 2024.2 release announcement is this one - https://github.com/datalust/seq-tickets/issues/2136?ref=blog.datalust.co - but there's not much more info there than what we've collected in this thread.

Just regarding the downgrade to 2024.1, I'd recommend trying to move forward again as soon as possible, as differences in the storage/index format are likely to cause issues on 2024.1 if any indexes were updated by the new version.

If using the new provider name doesn't resolve it, resetting authentication using an init script and reconfiguring Entra ID auth on 2024.3 should sort things out:

https://docs.datalust.co/docs/recovering-from-broken-authentication-settings

Hope this helps,
Nick

**Reply by Mike-E-angelo** [reply]
Thank you for the continued support Nick!  It is appreciated.  Please note that I am testing this currently on new deployments, so downgrades are not occurring. 👍

I am using the following on `datalust/seq:latest` and it is still resulting in our `//` issue on a new install:

```
seqcli setting set -n=EntraIDAuthority -v="login.microsoftonline.com" -s $Address --apikey=$system
        seqcli setting set -n=EntraIDClientId -v="$application" -s $Address --apikey=$system
        seqcli setting set -n=EntraIDClientKey -v="$secret" -s $Address --apikey=$system
        seqcli setting set -n=EntraIDTenantId -v="$TenantId" -s $Address --apikey=$system
        seqcli setting set -n=AuthenticationProvider -v="Microsoft Entra ID" -s $Address --apikey=$system
```

Is there something else I am overlooking, by chance?

**Reply by nblumhardt** [reply]
Hi Mike; is it possible your seqcli version is behind?

On a fresh Seq container, I run:

```
seqcli setting set -n=EntraIDAuthority -v="login.microsoftonline.com"
seqcli setting set -n=EntraIDClientId -v="application"
seqcli setting set -n=EntraIDClientKey -v="secret"
seqcli setting set -n=EntraIDTenantId -v="TenantId" 
seqcli setting set -n=AuthenticationProvider -v="Microsoft Entra ID"
seqcli setting set -n=IsAuthenticationEnabled -v="True"
```

And as expected I get `TenantId` in the URL:

```
System.InvalidOperationException: IDX20803: Unable to obtain configuration from: 'https://login.microsoftonline.com/TenantId/v2.0/.well-known/openid-configuration'. Will retry at '05/24/2024 21:37:30 +00:00'. Exception: 'System.IO.IOException: IDX20807: Unable to retrieve...
```

What I do notice in your example is that your tenant id variable is cased differently from the others; this might work in PowerShell, but if provisioning is happening in bash you might have a bug:

```
bash-3.2$ x=5
bash-3.2$ echo $x
5
bash-3.2$ echo $X

bash-3.2$ 
```

**Reply by Mike-E-angelo** [reply]
Thank you for the further consideration @nblumhardt.  I do use PowerShell for the deployment so we can rule that out.  Is the `IsAuthenticationEnabled` required for the latest release?  I can try adding that and see if that works.  At the moment I have to pin to `datalust/seq:2024.1.11028` as that is the only way I can get this to work.  I have some time today and will further investigate. 👍

**Reply by Mike-E-angelo** [reply, marked as answer]
Alright @nblumhardt I have a clue here.  I tried to output the value seqcli (version `2024.3.873`) thinks is in the `EntraIdTenantId`, but when I attempt to do so using the following commands this is what I get:

```
> seqcli setting show -n=EntraIDTenantId --server=asdf
The setting name EntraIDTenantId was not recognized; run the `seqcli setting names` command to see supported settings.

> seqcli setting show -n=EntraIDTenantId -s http://asdf      
The setting name EntraIDTenantId was not recognized; run the `seqcli setting names` command to see supported settings.
```

Note the assigned server which is invalid and doesn't exist.  It seems that it's not using this value and is instead connecting to my local server (which currently has a previous version installed).

EDIT;  My assumption was that `dotnet tool install --global seqcli` installed its own copy of the latest API for its own uses, but it appears to use whatever is installed locally instead. 🤔  **Installing the latest Seq server locally fixed the issue and everything works again.** 😌
